Modules/_zstd/_zstdmodule.c

```diff
@@ -0,0 +1,968 @@
+                 ZSTD_versionString(), 8*(int)sizeof(Py_ssize_t));
+}
+
+/* -------------------------
+     Train dictionary code
+   ------------------------- */
+
+/*[clinic input]
+_zstd._train_dict
+
+        samples_bytes: PyBytesObject
+            Concatenation of samples.
+        samples_size_list: object(subclass_of='&PyList_Type')
+            List of samples' sizes.
+        dict_size: Py_ssize_t
+            The size of the dictionary.
+        /
+
+Internal function, train a zstd dictionary on sample data.
+[clinic start generated code]*/
+
+static PyObject *
+_zstd__train_dict_impl(PyObject *module, PyBytesObject *samples_bytes,
+                       PyObject *samples_size_list, Py_ssize_t dict_size)
+/*[clinic end generated code: output=ee53c34c8f77886b input=fdb00ebe474d1df9]*/
+{
+    Py_ssize_t chunks_number;
+    size_t *chunk_sizes = NULL;
+    PyObject *dst_dict_bytes = NULL;
+    size_t zstd_ret;
+    Py_ssize_t sizes_sum;
+    Py_ssize_t i;
+
+    /* Check arguments */
+    if (dict_size <= 0) {
+        PyErr_SetString(PyExc_ValueError, "dict_size argument should be positive number.");
+        return NULL;
+    }
+
+    chunks_number = Py_SIZE(samples_size_list);
+    if ((size_t) chunks_number > UINT32_MAX) {
+        PyErr_SetString(PyExc_ValueError,
+                        "The number of samples should <= UINT32_MAX.");
+        return NULL;
+    }
+
+    /* Prepare chunk_sizes */
+    chunk_sizes = PyMem_Malloc(chunks_number * sizeof(size_t));
+    if (chunk_sizes == NULL) {
+        PyErr_NoMemory();
+        goto error;
+    }
+
+    sizes_sum = 0;
+    for (i = 0; i < chunks_number; i++) {
+        PyObject *size = PyList_GET_ITEM(samples_size_list, i);
+        chunk_sizes[i] = PyLong_AsSize_t(size);
+        if (chunk_sizes[i] == (size_t)-1 && PyErr_Occurred()) {
+            PyErr_SetString(PyExc_ValueError,
+                            "Items in samples_size_list should be an int "
+                            "object, with a size_t value.");
+            goto error;
+        }
+        sizes_sum += chunk_sizes[i];
+    }
+
+    if (sizes_sum != Py_SIZE(samples_bytes)) {
+        PyErr_SetString(PyExc_ValueError,
+                        "The samples size list doesn't match the concatenation's size.");
+        goto error;
+    }
+
+    /* Allocate dict buffer */
+    dst_dict_bytes = PyBytes_FromStringAndSize(NULL, dict_size);
+    if (dst_dict_bytes == NULL) {
+        goto error;
+    }
+
+    /* Train the dictionary */
+    Py_BEGIN_ALLOW_THREADS
+    zstd_ret = ZDICT_trainFromBuffer(PyBytes_AS_STRING(dst_dict_bytes), dict_size,
+                                     PyBytes_AS_STRING(samples_bytes),
+                                     chunk_sizes, (uint32_t)chunks_number);
+    Py_END_ALLOW_THREADS
+
+    /* Check zstd dict error */
+    if (ZDICT_isError(zstd_ret)) {
+        _zstd_state* const _module_state = PyModule_GetState(module);
```

Comment:
if the module state is NULL, no exception is set. We should abort here. This is why it's preferrable to have a function, say, `get_module_state(module)` which returns the module state but with an assert. Something like:

```c
static _zstd_state *get_state(PyObject *m) {
    void *s = PyModule_GetState(m);
    assert(s != NULL);
    return (_zstd_state *)s;
}
```

Reply:
I pushed a commit adding a module get state function right after your review 😁